File: src/lib/merge-driver.ts
#!/usr/bin/env bun
/**
 * Git merge driver for Automerge .automerge files.
 *
 * Registered via .gitattributes + git config. When git encounters a merge
 * conflict on a .automerge file, it calls this script instead of its
 * default binary conflict handler.
 *
 * Usage (called by git automatically):
 *   bun merge-driver.ts <base> <ours> <theirs>
 *
 * - Loads all three versions as Automerge documents
 * - Merges ours + theirs (Automerge merge is conflict-free)
 * - Writes the result to <ours> (git convention)
 * - Exits 0 on success (tells git the merge succeeded)
 */

import * as Automerge from "@automerge/automerge";
import type { Project } from "./schema.js";

const [, , basePath, oursPath, theirsPath] = process.argv;

if (!basePath || !oursPath || !theirsPath) {
  console.error("Usage: merge-driver.ts <base> <ours> <theirs>");
  process.exit(1);
}

async function loadFile(path: string): Promise<Automerge.Doc<Project>> {
  const buffer = await Bun.file(path).arrayBuffer();
  return Automerge.load<Project>(new Uint8Array(buffer));
}

try {
  const ours = await loadFile(oursPath);
  const theirs = await loadFile(theirsPath);

  // Automerge.merge is conflict-free — concurrent changes combine cleanly
  const merged = Automerge.merge(ours, theirs);

  // Write merged result back to "ours" path (git convention)
  const binary = Automerge.save(merged);
  await Bun.write(oursPath, binary);

  process.exit(0);
} catch (err) {
  console.error("Automerge merge driver failed:", err);
  process.exit(1);
}

File: src/rust/crates/agt-lib/src/merge_driver.rs
//! Git merge driver for Automerge .automerge files.
//!
//! When git encounters a merge conflict on a .automerge file, it calls
//! this instead of its default binary conflict handler.

use anyhow::{Context, Result};
use automerge::AutoCommit;
use std::path::Path;

/// Perform a three-way merge of Automerge documents.
/// Writes the merged result to `ours_path` (git convention).
pub fn merge_files(
    _base_path: &Path,
    ours_path: &Path,
    theirs_path: &Path,
) -> Result<()> {
    let ours_bytes = std::fs::read(ours_path)
        .with_context(|| format!("failed to read {}", ours_path.display()))?;
    let theirs_bytes = std::fs::read(theirs_path)
        .with_context(|| format!("failed to read {}", theirs_path.display()))?;

    let mut ours = AutoCommit::load(&ours_bytes)
        .with_context(|| "failed to load ours")?;
    let mut theirs = AutoCommit::load(&theirs_bytes)
        .with_context(|| "failed to load theirs")?;

    // Automerge merge is conflict-free — concurrent changes combine cleanly
    ours.merge(&mut theirs)
        .with_context(|| "automerge merge failed")?;

    let binary = ours.save();
    std::fs::write(ours_path, binary)
        .with_context(|| format!("failed to write merged result to {}", ours_path.display()))?;

    Ok(())
}
